src/mlpack/tests/main_tests/range_search_test.cpp

```diff
@@ -0,0 +1,292 @@
+  CheckMatrices(distances, distance_val);
+  cout<<" 4.Results Verified"<<endl;
+
+  cout<<"Passed Synthetic Test 1"<<endl<<endl;
+}
+
+BOOST_AUTO_TEST_CASE(ParameterTesting)
+{
+  cout<<"Synthetic Test 2"<<endl;
+  arma::mat query_data={{5, 3, 1},{4, 2, 4},{3, 1, 7}};
+  arma::mat x={{0, 3, 3, 4, 3, 1},{4, 4, 4, 5, 5, 2},{0, 1, 2, 2, 3, 3}};
+  vector<std::vector<double>> distance_val={
+                {2.82843, 2.23607, 1.73205, 2.23607, 4.47214},
+                {3.74166, 2, 2.23607, 3.31662, 3.60555, 2.82843},
+                {4.58258, 4.47214}};
+  vector<std::vector<size_t>> neighbor_val = {{1,2,3,4,5},{0,1,2,3,4,5},{4,5}};
+  std::string distance_file = "distances.csv";
+  std::string neighbors_file = "neighbors.csv";
+  double min_v = 0, max_v = 5;
+
+  SetInputParam("query", query_data);
+  SetInputParam("reference", std::move(x));
+  SetInputParam("min", min_v);
+  SetInputParam("max", max_v);
+  SetInputParam("distances_file", distance_file);
+  SetInputParam("neighbors_file", neighbors_file);
+  cout<<" 1.Parameters Set"<<endl;
+
+  mlpackMain();
+
+  cout<<" 2.Model Created"<<endl;
+  math::Range r(min_v, max_v);
+  vector<vector<size_t>> neighbors;
+  vector<vector<double>> distances;
+  CLI::GetParam<RSModel*>("output_model")->Search(std::move(query_data), r,
+                                                  neighbors, distances);
+  cout<<" 3.Search with Query Executed"<<endl;
+
+  CheckMatrices(neighbors, neighbor_val, 1e-5);
+  CheckMatrices(distances, distance_val);
+
+  cout<<" 4.Results Verified"<<endl;
+
+  cout<<"Passed Synthetic Test 2"<<endl<<endl;
+}
+
+BOOST_AUTO_TEST_CASE(ModelCheck)
+{//check if an output model can be used again for a different usage
+  cout<<"Model Reusability Test"<<endl;
+  arma::mat input_data, query_data;
+  double min_v = 0, max_v = 3;
+  std::string distance_file = "distances.csv";
+  std::string neighbor_file = "neighbors.csv";
+
+  if (!data::Load("iris.csv", input_data))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+  if (!data::Load("iris_test.csv", query_data))
+    BOOST_FAIL("Unable to load dataset iris_test.csv!");
+
+  SetInputParam("reference", std::move(input_data));
+  SetInputParam("min", min_v);
+  SetInputParam("max", max_v);
+  SetInputParam("distances_file", distance_file);
+  SetInputParam("neighbors_file", neighbor_file);
+  cout<<" 1.Parameters Set"<<endl;
+
+  mlpackMain();
+
+  RSModel* output_model = std::move(CLI::GetParam<RSModel*>("output_model"));
+  CLI::GetSingleton().Parameters()["reference"].wasPassed=false;
+  cout<<" 2.Model Created and copied"<<endl;
+
+  SetInputParam("input_model", std::move(output_model));
+  SetInputParam("query", std::move(query_data));
+
+  mlpackMain();
+
+  cout<<" 3.Model Reinserted"<<endl;
+  if (!( output_model == CLI::GetParam<RSModel*>("output_model") ))
+  {
+    BOOST_FAIL("Models are not Equal");
+  }
+  else
+  {
+    cout<<"Model Checking Test Passed"<<endl<<endl;
+  }
```

Comment:
It would be worth also testing that the results of the range search are the same for both the first and second runs of `mlpackMain()`.